### PR TITLE
Added comments,tests and neccesary changes

### PR DESCRIPTION
**Pull request by kpulkit29**

### CODE REVIEW

**Comment by kpulkit29** [comment]
Please review it

**Comment by kpulkit29** [comment]
kindly note the following things -:
The tokenVesting.sol imports erc20.sol which is an abstract contract.As I am written test and deployed on local net truffle throws error when deploying abstract contract.So I have changed the import statement to import "openzeppelin-solidity/contracts/token/ERC20/MintableToken.sol";
Only then was I able to mint tokens  execute the functions

**Comment by kpulkit29** [comment]
All the tests are passing

File: test/1_TokenVesting.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  })
+    
+it("should  not be able to claim funds before the cliff",function(){
+    return tokenVest.claimTokens().then(res=>{
+
+    }).catch(err=>{
+        assert.isDefined(err,"no error");
+    });
+})
+it("owner should be able to kill vesting",function(){
```
Comment (adamskrodzki) [inline test/1_TokenVesting.js:66]: Where are any tests checking that vesting actually works and sum releassed is as expected?

**Comment by kpulkit29** [comment]
The claimToken() function will not work until the cliff time has passed ,in
order to test that I will have to deploy the contracts again different
params

**Comment by adamskrodzki** [comment]
No you do not need do that, here is correct approach

https://medium.com/level-k/testing-smart-contracts-with-truffle-7849b3d9961

File: contracts/TokenVesting.sol [hunk 2]
```diff
@@ -6,20 +6,40 @@ import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 
 contract TokenVesting is Ownable {
   using SafeMath for uint;
-
+  
+  //erc20 conpliant token address
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:10]: please change for `ERC20 token address`

File: contracts/TokenVesting.sol [hunk 3]
```diff
@@ -6,20 +6,40 @@ import "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+  
+  //Event emitted when the tokens are added to the contract address for transfer to receiver in the future
   event VestingFunded(uint256 totalTokens);
+  //Event emitted when token claimed by investor
   event TokensClaimed(uint256 tokensClaimed);
+  //event emitted when the remaining tokens are transfered back to the owner
   event VestingKilled();
 
+  /**
+   *@dev contract contructor for initialization
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:35]: just `contract constructor`

File: contracts/TokenVesting.sol [hunk 4]
```diff
@@ -6,20 +6,40 @@ import "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+   *@param _token address of erc20 token
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:36]: ERC20

File: contracts/TokenVesting.sol [hunk 5]
```diff
@@ -35,7 +55,12 @@ contract TokenVesting is Ownable {
     totalPeriods = _totalPeriods;
     timePerPeriod = _timePerPeriod;
   }
-
+ 
+ /**
+  *@dev `fundVesting` allows to add token to the contract address for transfers
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:60]: add that can be executed only once please

File: contracts/TokenVesting.sol [hunk 6]
```diff
@@ -61,7 +90,10 @@ contract TokenVesting is Ownable {
 
     emit TokensClaimed(tokensToClaim);
   }
-
+ /**
+  *@dev `killVesting` to finish the vesting process
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:94]: Please add that it can be called only by the contract owner

File: contracts/TokenVesting.sol [hunk 7]
```diff
@@ -45,6 +70,10 @@ contract TokenVesting is Ownable {
     emit VestingFunded(_totalTokens);
   }
 
+ /**
+  *@dev `claimTokens` to claim tokens called by receiver
+  *Emit `TokensClaimed` event
+  */
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:76]: Please add taht can be called **only** by receiver

File: contracts/TokenVesting.sol [hunk 8]
```diff
@@ -1,6 +1,6 @@
 pragma solidity ^0.4.24;
 
-import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/MintableToken.sol";
```
Comment (AugustoL) [inline contracts/TokenVesting.sol:3]: Please remove this change. It should be ERC20 not mintable

File: test/1_TokenVesting.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    assert.isUndefined(res,"no error");
+  });
+})
+it("should be able to fund vesting when contract is not approved for such amount of tokens",function(){
+    return tokenVest.fundVesting(1000).then((res)=>{
+    }).catch(function(err){
+      assert.isDefined(err,"there is err");
+    });
+  })
+it("should be not able to fund vesting again after doing it once",function(){
```
Comment (AugustoL) [inline test/1_TokenVesting.js]: should not*

File: test/1_TokenVesting.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    return tokenVest.fundVesting(100).then((res)=>{
+    }).catch(function(err){
+      assert.isDefined(err,err);
+    });
+  })
+    
+it("should  not be able to claim funds before the cliff",function(){
```
Comment (AugustoL) [inline test/1_TokenVesting.js]: double space between should and not

File: test/1_TokenVesting.js (hunk 1 above)
Comment (AugustoL) [reply test/1_TokenVesting.js:66]: Yes, seems that those tests are missing.

Comment (AugustoL) [reply test/1_TokenVesting.js:66]: ping @kpulkit29

**Review by AugustoL — CHANGES_REQUESTED** [review]
@kpulkit29 Looks good so far, some tests are missing. Leave the import using ERC20 and if you need a token to test just deploy a StandardToken form zeppelin in beforeEach test.
Also, can you write the tests using async/await?

Thanks for your work !

**Comment by AugustoL** [comment]
@kpulkit29 there is still a lot of changes requested missing and the tests are not using increase_time function. Can you finish it? If not I will, let me know.

**Comment by kpulkit29** [comment]
I don't think there is any increase_time function in the vesting contract.

**Comment by kpulkit29** [comment]
Except for the claiming token test after cliff  test I have covered all the
tests and added all comments.This is what was mentioned in the bounty.I
request you to review my work and payout the bounty.
Thanks,
Pulkit

**Comment by AugustoL** [comment]
@kpulkit29 to test the claimTokens you need to increase in the testrpc or ganache. Take a look at this helpers for test that we use in the lif-token repository https://github.com/windingtree/lif-token/tree/master/test/helpers

The tests should be in the test folder and cover all the contract functions and conditions.

You are not cheking all conditions.

Your work was already reviewed but is not approved, I requested changes, once you do it you will get your bounty.

**Comment by adamskrodzki** [comment]
I s this bounty still opened? I do not want to work for nothing, but I can do this today

**Comment by AugustoL** [comment]
@adamskrodzki still opened.

**Pull request closed without merging**